### PR TITLE
Unhealthy channel is not offered back to the pool.

### PR DESCRIPTION
**Pull request by ioanbsu**
Motivation:
When releasing unhealthy channel back to a pool we don't have to offer it since on acquire it will be discarded anyways.
 Also checking healthiness at release is a good idea so we don't end up having tons of unhealthy channels in the pool(unless they became unhealthy after being offered)

Modifications:
private SimpleChannelPool.offerIfHealthy() method added that is called from SimpleChannelPool.doReleaseChannel(). SimpleChannelPool.offerIfHealthy() offers channel back to pool only if channel is healthy.
Otherwise it throws setFailure exception to the promise.

 Result:
The pool is now much cleaner and not spammed with unhealthy channels.

See https://github.com/netty/netty/issues/4077 for more details.

### CODE REVIEW

**Comment by ioanbsu** [comment]
@normanmaurer , hm, again this test, keeps failing. I don't believe it's my changes.

**Comment by normanmaurer** [comment]
will check tomorrow

**Comment by Scottmitch** [comment]
@ioanbsu - Build failure is not related to this PR

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 1]
```diff
@@ -230,11 +232,16 @@ private void doReleaseChannel(Channel channel, Promise<Void> promise) {
                          promise);
         } else {
             try {
-                if (offerChannel(channel)) {
-                    handler.channelReleased(channel);
-                    promise.setSuccess(null);
+                final Future<Boolean> f = healthCheck.isHealthy(channel);
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: As discussed in https://github.com/netty/netty/issues/4077#issuecomment-130461684. Perhaps we should make this check configurable (some boolean flag to indicate we should check health when putting back into pool).

Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: @Scottmitch sure, will do. What do you think the default behavior should be? Check health on release by default or do not check by default?

Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: My vote is default should check the health. If this hurts performance in specific use cases it can be switched off. Our default health checker `ChannelHealthChecker.ACTIVE` shouldn't involve any network traffic so should be OK. @normanmaurer - WDYT?

**Comment by ioanbsu** [comment]
hm, is there a way to see what exact checkstyle rule failed?
Also can I get checkstyle.xml file so I can run checks before pushing my code?

**Comment by normanmaurer** [comment]
@ioanbsu yes:
https://garage.netty.io/teamcity/viewLog.html?buildId=325&buildTypeId=pulls_netty&tab=buildLog#_focus=7997

**Comment by ioanbsu** [comment]
@Scottmitch, @normanmaurer I'me done with changes. lmk what you think.

Also I've noticed there are couple typos in FixedChannelPool such as "numnber", "healty" "connetion" and etc. lmk if it's fine if I fix it in this pull request as well.

**Comment by trustin** [comment]
You can just try to build from command line before pushing. :-) You can do 'mvn validate' to perform checkstyle tests only.

Also, the build output from the CI server also contains where the
checkstyle problems are. Please try to expand the build log just in case
it is collapsed.

**Comment by ioanbsu** [comment]
Thank you @trustin, Yeah with @normanmaurer 's help I've found it was in teamcity's logs.
I was wondering if I can get this file "io/netty/checkstyle.xml". I see the maven configured to use this file but I don't see it in repository. It's also a bit confusing how would ```mvn validate``` work locally if there is now checkstyle.xml that it can read.
I'd like to configure IDEA so it highlights me right away in the code if there anything wrong.

**Comment by trustin** [comment]
Ah, it's in the netty-build project:

https://github.com/netty/netty-build

It uses some custom rules, so I'm not sure it will work properly with
IDE.

**Comment by normanmaurer** [comment]
@ioanbsu why not only always offer  healty channels ? I don't think we need an extra method here.

**Comment by ioanbsu** [comment]
@normanmaurer only offer healthy channels - that was my original implementation. But then @Scottmitch  and you(if I got it right) suggested to give an option for developers to choose if they want to offer any channel or healthy only.

The motivation, as I got it, was that we going to check health on acquire anyways so we don't necessarily need to check health on offer because check channel health may be expensive operation.

**Comment by normanmaurer** [comment]
@ioanbsu yeah but I think add a constructor param is good enough.

**Comment by ioanbsu** [comment]
Sure, I can do it through constructor.

Right now we have two constructors in SimpleChannelPool: 
``` public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler)```
```public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler, ChannelHealthChecker healthCheck)```
Should I just update the second one to:
```public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler, ChannelHealthChecker healthCheck)``` ->
```public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler, ChannelHealthChecker healthCheck, boolean offerHealthyOnly)``` ?

Also for FixedChannelPooll we have 3 constructors:
```public FixedChannelPool(Bootstrap bootstrap,  ChannelPoolHandler handler, int maxConnections)```
```public FixedChannelPool(Bootstrap bootstrap,  ChannelPoolHandler handler, int maxConnections, int maxPendingAcquires)```
```public FixedChannelPool(Bootstrap bootstrap, ChannelPoolHandler handler,           ChannelHealthChecker healthCheck, AcquireTimeoutAction action, final long acquireTimeoutMillis,                            int maxConnections, int maxPendingAcquires)```

Should I just update: 
```public FixedChannelPool(Bootstrap bootstrap, ChannelPoolHandler handler,           ChannelHealthChecker healthCheck, AcquireTimeoutAction action, final long acquireTimeoutMillis,                            int maxConnections, int maxPendingAcquires)``` ->
```public FixedChannelPool(Bootstrap bootstrap, ChannelPoolHandler handler,           ChannelHealthChecker healthCheck, AcquireTimeoutAction action, final long acquireTimeoutMillis,                            int maxConnections, int maxPendingAcquires, boolean offerHealthyOnly )``` ?

Also I'm envisioning boolean offerHealthyOnly  be true by default. Please let me know if you ok with that.

File: transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java [hunk 2]
```diff
@@ -142,4 +149,95 @@ protected boolean offerChannel(Channel ch) {
...
+        ChannelPoolHandler handler = new CountingChannelPoolHandler();
+        ChannelPool pool = new SimpleChannelPool(cb, handler);
+        Channel channel1 = pool.acquire().syncUninterruptibly().getNow();
+        pool.release(channel1).syncUninterruptibly();
+        Channel channel2 = pool.acquire().syncUninterruptibly().getNow();
+        //first check that when returned healthy then it actually offered back to the pool.
+        assertSame(channel1, channel2);
+
+        expectedException.expect(IllegalStateException.class);
+        expectedException.expectMessage("Channel is unhealthy not offering it back to pool");
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java]: I'm not sure we need to verify / require this text is in the message.

Also do we need the `ExpectedException` `@Rule`?  Is this not something we can catch in normal code flow? I think we typically catch expected exceptions and `fail()` if we don't get the exception.

**Comment by Scottmitch** [comment]
@ioanbsu - Just add a new constructor (where needed) with an extra argument `boolean offerHealthyOnly` instead of modifying existing constructor signatures, and call the new constructor from the existing constructor.

File: transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java (hunk 2 above)
Comment (ioanbsu) [inline transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java]: So that I know for sure that expected exception(not any random) had been thrown.

**Comment by ioanbsu** [comment]
@Scottmitch  sure, will do.

File: transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java (hunk 2 above)
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java]: @ioanbsu - Understood, but I don't think we need to couple the tests so tightly that the message in the exception is verified. I think just verifying the correct type of exception is thrown at the correct time should be sufficient.

Comment (ioanbsu) [inline transport/src/test/java/io/netty/channel/pool/SimpleChannelPoolTest.java]: @Scottmitch, got it. removed message validation.

**Comment by ioanbsu** [comment]
@normanmaurer  @Scottmitch @trustin I believe the test that failed has nothing to do with my changes.
I applied all the changes that you requested. Please let me know if there anything else left.
Also please let me know if you'd like me to squash my changes.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 3]
```diff
@@ -43,6 +43,8 @@
 public class SimpleChannelPool implements ChannelPool {
     private static final AttributeKey<SimpleChannelPool> POOL_KEY = AttributeKey.newInstance("channelPool");
     private static final IllegalStateException FULL_EXCEPTION = new IllegalStateException("ChannelPool full");
+    private static final IllegalStateException UNHEALTHY_NON_OFFERED_TO_POOL =
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: You also need to empty the stack trace as done for "FULL_EXCEPTION" below.

**Comment by normanmaurer** [comment]
@ioanbsu just one more thing to fix. After this please squash and I will cherry-pick

**Comment by ioanbsu** [comment]
@normanmaurer done and squashed.

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java [hunk 4]
```diff
@@ -121,7 +121,33 @@ public FixedChannelPool(Bootstrap bootstrap,
...
+     * @param action                the {@link AcquireTimeoutAction} to use or {@code null} if non should be used.
+     *                              In this case {@param acquireTimeoutMillis} must be {@code -1}.
+     * @param acquireTimeoutMillis  the time (in milliseconds) after which an pending acquire must complete or
+     *                              the {@link AcquireTimeoutAction} takes place.
+     * @param maxConnections        the numnber of maximal active connections, once this is reached new tries to
+     *                              acquire a {@link Channel} will be delayed until a connection is returned to the
+     *                              pool again.
+     * @param maxPendingAcquires    the maximum number of pending acquires. Once this is exceed acquire tries will
+     *                              be failed.
+     * @param offerHealthyOnly      will check channel health before offering back if this parameter set to true.
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: `true` -> `{@code true}`.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 5]
```diff
@@ -67,11 +71,26 @@ public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler)
...
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param bootstrap         the {@link Bootstrap} that is used for connections
+     * @param handler           the {@link ChannelPoolHandler} that will be notified for the different pool actions
+     * @param healthCheck       the {@link ChannelHealthChecker} that will be used to check if a {@link Channel} is
+     *                          still healthy when obtain from the {@link ChannelPool}
+     * @param offerHealthyOnly  will check channel health before offering back if this parameter set to true.
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: `true` -> `{@code true}`.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 6]
```diff
@@ -219,7 +238,7 @@ public void run() {
         return promise;
     }
 
-    private void doReleaseChannel(Channel channel, Promise<Void> promise) {
+    private void doReleaseChannel(final Channel channel, final Promise<Void> promise, final boolean offerHealthyOnly) {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Is it necessary to have this as an argument? Seems like we are just passing in the member variable when ever this is called.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 7]
```diff
@@ -43,13 +43,17 @@
...
+            new IllegalStateException("Channel is unhealthy not offering it back to pool");
     static {
         FULL_EXCEPTION.setStackTrace(EmptyArrays.EMPTY_STACK_TRACE);
+        UNHEALTHY_NON_OFFERED_TO_POOL.setStackTrace(EmptyArrays.EMPTY_STACK_TRACE);
     }
     private final Deque<Channel> deque = PlatformDependent.newConcurrentDeque();
     private final ChannelPoolHandler handler;
     private final ChannelHealthChecker healthCheck;
     private final Bootstrap bootstrap;
+    private final boolean offerHealthyOnly;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Consider renaming to `releaseHeathCheck`....and also the same for the publicly visible in constructors.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 8]
```diff
@@ -67,11 +71,26 @@ public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler)
...
+     * Creates a new instance.
+     *
+     * @param bootstrap         the {@link Bootstrap} that is used for connections
+     * @param handler           the {@link ChannelPoolHandler} that will be notified for the different pool actions
+     * @param healthCheck       the {@link ChannelHealthChecker} that will be used to check if a {@link Channel} is
+     *                          still healthy when obtain from the {@link ChannelPool}
+     * @param offerHealthyOnly  will check channel health before offering back if this parameter set to true.
+     */
+    public SimpleChannelPool(Bootstrap bootstrap, final ChannelPoolHandler handler, ChannelHealthChecker healthCheck,
+                             final boolean offerHealthyOnly) {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: remove `final`.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 9]
```diff
@@ -183,9 +202,9 @@ private void notifyHealthCheck(Future<Boolean> future, Channel ch, Promise<Chann
     }
 
     /**
-     * Bootstrap a new {@link Channel}. The default implementation uses {@link Bootstrap#connect()},
-     * sub-classes may override this.
-     *
+     * Bootstrap a new {@link Channel}. The default implementation uses {@link Bootstrap#connect()}, sub-classes may
+     * override this.
+     * <p/>
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: should this just be `<p>`, or do you want to wrap the above in `<p>...</p>`?

**Comment by Scottmitch** [comment]
@ioanbsu - The test failure is unrelated to your changes.

I just had a few more comments to resolve before cherry-pick.

**Comment by ioanbsu** [comment]
@Scottmitch omg, :) I just squashed:). 
Ok ,let me address your comments.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java (hunk 6 above)
Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: You're right. will fix.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java (hunk 8 above)
Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Why?

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java (hunk 9 above)
Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: This is not part of my changes actually. This part got auto-reformatted by IDEA based on the rules that are used in netty project. I can certainly change it to <p>

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java (hunk 8 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: just a nit, but it is not needed.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 10]
```diff
@@ -219,7 +238,7 @@ public void run() {
         return promise;
     }
 
-    private void doReleaseChannel(Channel channel, Promise<Void> promise) {
+    private void doReleaseChannel(final Channel channel, final Promise<Void> promise) {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Why this change?

Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: This is general code style practice that I try to follow in my projects. It helps to prevent possible bugs. If there's no need to change the value of a variable and someone does, the IDE and/or compiler will provide a warning.
It may also be a helping hint to compiler to perform some optimizations for the code.

I don't mind removing it if in netty projects you don't have such rules.

Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Lets revert for this PR. I understand the rational, and do not necessarily disagree, but lets keep the set of changes focused on what is required to resolve the issue.  If need be we can open a followup PR focused on using `final` more consistently.

Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Sounds good.

Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Done.

**Comment by Scottmitch** [comment]
@ioanbsu - Thanks for your patience! Can you squash again and I will cherry-pick?

**Comment by ioanbsu** [comment]
Sure. np. I was waiting for final call before squashing.

**Comment by ioanbsu** [comment]
@Scottmitch done.

**Comment by ioanbsu** [comment]
btw any reason not to include it in 4.1.0.Beta6? I'm using this version in my project and it would be delightful is I don't need to override these classes locally.

**Comment by Scottmitch** [comment]
Thanks again @ioanbsu! Good work!

Cherry-picked 4.0 (b46e070) 4.1 (08b73bf) master (144f3ec)

**Comment by ioanbsu** [comment]
Ah nevermind, I see it cherry-picked to 4.1:)
